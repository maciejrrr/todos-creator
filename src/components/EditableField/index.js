import React, { Component } from 'react';
import styled from 'styled-components';

import { colors, metrics } from '../../theme';

const Name = styled.a`
  color: ${colors.darkGrayText};
  font-size: ${metrics.fontSize.lg}px;
  word-wrap: break-word;
  margin-bottom: ${metrics.smallMargin}px;
  :hover {
    color: ${colors.darkGrayText};
  }
`;

const Input = styled.input`
  color: ${colors.darkGrayText};
  height: ${metrics.inputHeight}px;
  padding: ${metrics.smallMargin}px;
  margin-bottom: ${metrics.smallMargin}px;
  width: 100%;
  border: 0;
  outline: 0;
`;

class EditableField extends Component {
  state = {
    editable: false,
    text: '',
  };

  handleFieldClick = () => {
    this.setState({ editable: true });
  };

  handleKeyPress = e => {
    const { edit, fieldId } = this.props;
    const { text } = this.state;
    if (e.key === 'Enter' && text) {
      e.preventDefault();
      edit({ text, id: fieldId });
      this.resetForm();
    }
  };

  handleChange = e => {
    this.setState({ text: e.target.value });
  };

  resetForm = () => {
    this.setState({ editable: false, text: '' });
  };

  render() {
    const { name } = this.props;
    const { editable, text } = this.state;
    let content = <Name onClick={this.handleFieldClick}>{name}</Name>;
    if (editable) {
      content = (
        <Input
          autoFocus
          value={text}
          onChange={this.handleChange}
          onKeyPress={this.handleKeyPress}
          onBlur={this.resetForm}
        />
      );
    }
    return content;
  }
}

export default EditableField;
