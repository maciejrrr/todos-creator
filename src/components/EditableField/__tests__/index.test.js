import EditableField from '../index';

describe('EditableField component', () => {
  let props;
  beforeEach(() => {
    props = {
      fieldId: 1,
      edit: jest.fn(),
      name: 'test field',
    };
  });

  describe('snapshots', () => {
    describe('initial state', () => {
      it('renders field name', () => {
        const tree = renderer.create(<EditableField {...props} />).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });

    describe('initial state', () => {
      it('renders field name', () => {
        const component = renderer.create(<EditableField {...props} />);
        component.getInstance().handleFieldClick();
        expect(component.toJSON()).toMatchSnapshot();
      });
    });
  });

  describe('handleFieldClick', () => {
    it('changes editable state', () => {
      const component = shallow(<EditableField {...props} />);
      component.instance().handleFieldClick();
      expect(component.state().editable).toBe(true);
    });
  });

  describe('handleKeyPress', () => {
    let e;
    let preventDefaultMock;
    let resetFormMock;
    beforeEach(() => {
      preventDefaultMock = jest.fn();
      resetFormMock = jest.fn();
    });

    describe('when submitting valid text', () => {
      let text;
      beforeEach(() => {
        const component = shallow(<EditableField {...props} />);
        e = { preventDefault: preventDefaultMock, key: 'Enter' };
        text = 'valid text';
        component.instance().resetForm = resetFormMock;
        component.setState({ text, editable: true });
        component.instance().handleKeyPress(e);
      });

      it('calls preventDefault', () => {
        expect(preventDefaultMock).toHaveBeenCalledTimes(1);
      });

      it('calls edit', () => {
        expect(props.edit).toBeCalledWith({
          text,
          id: props.fieldId,
        });
      });

      it('calls resetForm', () => {
        expect(resetFormMock).toHaveBeenCalledTimes(1);
      });
    });

    describe('not valid', () => {
      let text;
      const sharedBehavior = () => {
        it('does not call preventDefault', () => {
          expect(preventDefaultMock).not.toHaveBeenCalled();
        });

        it('does not call edit', () => {
          expect(props.edit).not.toHaveBeenCalled();
        });

        it('doed not call resetForm', () => {
          expect(resetFormMock).not.toHaveBeenCalled();
        });
      };

      describe('key pressed is not Enter', () => {
        beforeEach(() => {
          const component = shallow(<EditableField {...props} />);
          e = { preventDefault: preventDefaultMock };
          text = 'valid text';
          component.instance().resetForm = resetFormMock;
          component.setState({ text, editable: true });
          component.instance().handleKeyPress(e);
        });
        sharedBehavior();
      });

      describe('no text', () => {
        beforeEach(() => {
          const component = shallow(<EditableField {...props} />);
          e = { preventDefault: preventDefaultMock, key: 'Enter' };
          component.instance().resetForm = resetFormMock;
          component.setState({ editable: true });
          component.instance().handleKeyPress(e);
        });
        sharedBehavior();
      });
    });
  });

  describe('handleChange', () => {
    it('changes text state', () => {
      const component = shallow(<EditableField {...props} />);
      const text = 'test text';
      const e = { target: { value: text } };
      component.instance().handleChange(e);
      expect(component.state().text).toBe(text);
    });
  });

  describe('resetForm', () => {
    let component;
    beforeEach(() => {
      const text = 'test text';
      const e = { target: { value: text } };
      component = shallow(<EditableField {...props} />);
      component.setState({ editable: true, text });
      component.instance().resetForm(e);
    });

    it('changes text state', () => {
      expect(component.state().text).toBe('');
    });

    it('changes editable state', () => {
      expect(component.state().editable).toBe(false);
    });
  });
});
