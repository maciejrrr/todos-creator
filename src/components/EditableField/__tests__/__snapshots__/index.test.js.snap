// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableField component snapshots initial state renders field name 1`] = `
<a
  className="sc-bdVaJa eUzJrY"
  onClick={[Function]}
>
  test field
</a>
`;

exports[`EditableField component snapshots initial state renders field name 2`] = `
<input
  autoFocus={true}
  className="sc-bwzfXH bHPLBP"
  onBlur={[Function]}
  onChange={[Function]}
  onKeyPress={[Function]}
  value=""
/>
`;
